### ISSUE TITLE
Add Optional Fast Path for `clip_by_global_norm` (opt-in non-differentiable mode)

### ISSUE THREAD

**Issue by satyam-mishra-dev** [google-deepmind/optax#1452]
*Hi maintainers 👋, I’ve been exploring `optax/transforms/_clipping.py` and noticed an opportunity to add an opt-in performance improvement based on the existing TODO comment. Drafting this proposal for feedback before submitting a PR. I would love to contribute to deepmind*

The current implementation of `clip_by_global_norm` in `optax/transforms/_clipping.py` prioritizes preserving gradient flow, which is correct but slightly slower for standard use cases.
This proposal introduces an optional parameter to enable a faster, non-differentiable path for users who don’t require higher-order gradients.

---

### **Current Behavior**

Lines around 94–105 currently use a conditional implementation with `jax.lax.select`:

```python
trigger = jnp.squeeze(g_norm < max_norm)
chex.assert_shape(trigger, ())

def clip_fn(t):
    return jax.lax.select(trigger, t, (t / g_norm.astype(t.dtype)) * max_norm)

updates = jax.tree.map(clip_fn, updates)
```

A simpler version was previously commented out due to concerns about meta-gradient propagation:

```python
# g_norm = jnp.maximum(max_norm, g_norm)
# updates = jax.tree.map(lambda t: (t / g_norm) * max_norm, updates)
```

---

### **Proposed Enhancement**

Add an optional argument `allow_non_differentiable: bool = False` to `clip_by_global_norm`:

```python
def clip_by_global_norm(
    max_norm: float,
    *,
    allow_non_differentiable: bool = False
) -> base.GradientTransformation:
```

**Behavior:**

* When `allow_non_differentiable=False` (default):
  → Keeps current implementation (safe for meta-gradients).
* When `allow_non_differentiable=True`:
  → Uses the faster arithmetic-based implementation:

  ```python
  g_norm = jnp.maximum(max_norm, g_norm)
  updates = jax.tree.map(lambda t: (t / g_norm) * max_norm, updates)
  ```

---

### **Benefits**

1.  **Performance:** Faster clipping in training loops where differentiability through the optimizer is not required.
2.  **Flexibility:** Users can explicitly choose performance over meta-gradient safety.
3. **Safety:** Default behavior unchanged; compatible with all existing code.
4. **Cleanup:** Addresses TODO `b/163995078` with a controlled, user-configurable approach.

**Comment by rdyro** [comment]
Both versions are cache friendly, so, when compiled, I don't expect one to run faster than another. Both versions seem to be differentiable too.

To make any changes here, I'd like to understand what differences you're seeing in your application.

**Comment by satyam-mishra-dev** [comment]
I ran a small benchmark to double-check @rdyro 

Differentiable (current) → 0.02924 s/iter

Arithmetic (fast path) → 0.02830 s/iter
~3.3% faster on an 8192×8192 float32 tensor (JIT-compiled)

 XLA likely fuses most of the conditional overhead away, so the gain might be minor. Both versions also appear differentiable — the comment might’ve just been conservative.

This optimization probably only matters in heavy meta-learning loops where clipping runs thousands of times.
Happy to close this if you feel the change doesn’t add enough value — or I can run a meta-learning benchmark to confirm whether it compounds.

**Comment by rdyro** [comment]
Which hardware did you test?

It'd be very useful if you can find meta-learning cases where one, but not the other version produces NaNs. 

Unless we have strong evidence for needing this change, I'd prefer to keep this as is.

**Comment by satyam-mishra-dev** [comment]
I ran the benchmark on a MacBook M3 Pro (12-core CPU, 18 GB RAM) using JAX 0.4.33 with default Apple Metal backend.

I haven’t observed NaNs so far, but I’ll set up a few meta-learning inner-loop tests (e.g., MAML-style gradient updates) to see if any instability appears between the two implementations.

I’ll share results here once that’s done

**Comment by rdyro** [comment]
This is the same issue as: [https://github.com/google-deepmind/optax/issues/1535](https://github.com/google-deepmind/optax/issues/1535)

Closing with similar comments for now:
>Thanks for opening this issue, I don't see a strong enough justification for introducing this. The performance advantages are unlikely to be noticeable and it introduces unnecessary complexity both in code and for the end-user.
> 
> Please, reopen this issue if you are able to point to a real-world workflow where this alternative path improves performance (ideally not a toy example).
